### PR TITLE
fstab-generator: handle nofail mounts with timer-based trigger

### LINKED ISSUE #5991: User sessions not reliably stopped before filesystem unmounted on shutdown
### Submission type

  - [ XX ] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> …systemd-210, but still present upstream.

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

> …SLES12-SP1

### In case of bug report: Expected behaviour you didn't see

> …On shutdown, filesystems should not be unmounted until
all user session have been stopped.

### In case of bug report: Unexpected behaviour you saw

> …On shutdown, unmount of filesystems failed because some
user sessions were still running and were keeping the filesystem busy.
The network was then shutdown, still before the user session had completed.
The filesystems were NFS filesystem, and the user processes were
blocked writing to the NFS filesystem.  Now that the network was gone,
they blocked indefinitely.

### In case of bug report: Steps to reproduce the problem

> List an NFS filesystem in /etc/fstab as "noauto", then reboot, then mount
the filesysytem manually.

Then write a large file (e.g. use dd to create a few gigabytes) to the NFS filesystem.
This needs to be something that will take a while to complete.
The system needs enough memory to buffer enough data that flushing it
out will take tens of seconds.  This isn't an enormous requirement, but means
that it might be harder to duplicating on a tiny VM..
Once the write has started, "shutdown -r now".
This will 
- signal the 'dd' process, but it takes a while to die as it has to flush out the data
- try to unmount the NFS filesystem, and fail
- shut down the network
- try to sys_sync, and hang.

The root of the problem seems to be that fstab-generator doesn't create a Before=
dependency for noauto (and nofail) mounts. (in SLES, we also avoid the Before=
for "bg" nfs mounts, and that is what actually lead to us seeing the problem).
The "Before" dependency usually ensures that the mountpoints are unmounted *after*
some other stuff, which includes stopping the user processes.
"noauto" mounts don't get that ordering, so they are unmounted immediately.

I think that we need a new fs.target, possibly optional-fs.target.
fstab-generator would always create a Before= ordering, either against
an existing XX-fs.target, or against optional-fs.target.
The tricky bit, then, is getting optional-fs.target to start up, ever though some things
that are "Before" it might not have done so.
I *think* that this can be done be Want-ing optional-fs.target is a separate transaction from the transaction that mounts filesystem.  I suspect you understand the details of that better than me.

In any case, *.mounts that are not Before=XX-fs.target already need to be Before=optional-fs.target, and user sessions need to be After=optional-fs.target, without Wanting that target or mount points.

This idea seems to work in some simple testing, but I haven't pushed it very hard.

### PR DESCRIPTION
**Pull request by neilbrown**
fstab-generator omits the Before=fs-XXX.target dependency
on automount and nofail mounts.
This causes a problem at shutdown.  As the mounts are not "Before"
anything, they are initiated immediately, and might run while
services that use the mounts are still running.
This cause the mount to fail and never get unmounted.

This is particularly problematic for NFS the network can get shutdown
after these unmounts failed, and while services are still running.
If that happens, then any services (such as login sessions) using
the NFS mounts can block indefinitely - particularly if they had
written something to the NFS filesystem recently.  They cannot exit
until the data is written, and that cannot happen with the network down.

So we need that "Before=" dependency, but we don't want fs-XXX.target
to have to wait for the mount to complete.

For automounts, there is no difficulty.  The mount unit isn't activated
in the same transaction as the fs-XXX.target, so the ordering is not
effective.  We can re-instate the Before= dependency for automounts
without causing any problems.

For nofail mounts we need a little more work.  automounts set a good
example to follow.  If the mount unit is activated in a separate
transaction, the Before= dependency can be restored without causing
and delays are startup.  This is easily done with a timer unit.

So this patch:
 - restores the "Before=" dependency, so all mount units get it.
 - changes "nofail" handling to work much like "automount" except that
   a timer unit is created instead of an automount unit.  This timer
   unit has OnActiveSec=0 and triggers the mount unit.

Fixes #5991

### CODE REVIEW

**Comment by arvidjaar** [comment]
> changes "nofail" handling to work much like "automount" except that a timer unit is created instead of an automount unit. This timer unit has OnActiveSec=0 and triggers the mount unit.

Somehow I miss this part of patch.

**Comment by neilbrown** [comment]
> Somehow I miss this part of patch.

Yeah, sorry - it isn't there any more.  Someone pointed that that approach couldn't work because I was misunderstanding the independence of transaction.  So I came up with a new approach. 
Maybe I'm not using github properly, but is seems hard to keep conversations in one place and each pull request created a new conversation, which i would rather be part of the original issue..

**Comment by fbuihuu** [comment]
> Maybe I'm not using github properly, but is seems hard to keep conversations in one place and each pull request created a new conversation, which i would rather be part of the original issue..

I agree github makes it hard to follow.

I answered in the initial issue and I think this approach is the simplest one, we might want to document the new behavior though.

**Comment by fbuihuu** [comment]
@poettering could you have a look at Neil's changes ?

I think they make sense and remove the need of the asymmetric deps.

**Comment by poettering** [comment]
Hmm, this doesn't look right, sorry. The thing is that deps never go away after they are created. That means if you have a mount unit that is started once, then stopped it will have gained the dep, and that dep won't go away, hence a subsequent second start will be run with it in effect...

There are plans to track properly how each dep came to be, so that we can flush them out piecemeal, but that's not done yet, and not entirely trivial, but until that point I am not sure this can work...

File: src/core/mount.c [hunk 1]
```diff
@@ -1492,6 +1492,11 @@ static int mount_setup_existing_unit(
                 load_extras = true;
         }
 
+        if (!mount_is_extrinsic(MOUNT(u)) && !u->default_dependencies &&
```
Comment (fbuihuu) [inline src/core/mount.c]: I just found that the `!u->default_dependencies` should be inverted, i.e. the ordering dep should only be added if the mount unit has `DefaultDependencies=true`

**Comment by fbuihuu** [comment]
> Hmm, this doesn't look right, sorry. The thing is that deps never go away after they are created. That means if you have a mount unit that is started once, then stopped it will have gained the dep, and that dep won't go away, hence a subsequent second start will be run with it in effect...

Hmm I'm not sure to understand the problem here. The ordering dep patching is supposed to help during the boot process only but will have basically no effects after since at that point `local-fs.target` is always active.

**Comment by neilbrown** [comment]
I revised the commit to fixed the inversion of the test on ->default_dependencies.
I agree with @fbuihuu that after local-fs.target has been reached, future attempts to mount a unit with "Before=local-fs.target" will not be impacted, so adding the dependency which cannot be removed will not cause problems.

**Comment by poettering** [comment]
THe initial transaction should be repeatable, i.e. you should be able to "systemctl isolate emergency.target" and back to "systemctl isolate multi-user.target" and things should work. But if the the initial transaction modified tht dep tree, than it will be different from any repetition...

**Comment by neilbrown** [comment]
Thanks for the pointer to repeatability of the initial transaction.  I'll take your word for it that that is useful, but I have trouble seeing it as more important than the orderly normal startup and shutdown of the system - which might be seen as systemd's core responsibility.  It is the orderly shutdown that isn't working correctly here.  Until the removal of individual dependencies is implemented, you might need to just trigger a "daemon-reload" once emergency.target is reached.
The initial transaction *already* cannot be repeated if there are nofail network filesystems.
Commit a6d305f91d722e136c29222070efed5c5d8b120b adds Before=remote-fs.target to remote mounts when they are mounted.  Given the presence of that commit, there seems little justification for rejecting this pull request.

**Comment by gonzoleeman** [comment]
Until this is fixed, mounting and unmounting of remote filesystems, particularly at shutdown, is broken.

**Comment by brauner** [comment]
@neilbrown hey there. Didn't expect to see you here. I'm closing this since this has been unresolved for quite a few years. If this is still an issue that needs to be addressed please open a new pr.
